### ISSUE TITLE
group cycle timing inconsistent

### ISSUE THREAD

**Issue by guraltsev** [ultrabug/py3status#1777]
I have a group displaying some system information. The cycling, that according to the configuration is to be done every 2 seconds is however quite random. Sometimes a module stays there for 4-5 seconds, sometime it switches after 1sec. 

My bar is set to autohide so I am pressing the super key throughout this experiment.

Screencast gif attached 

![group-cycle](https://user-images.githubusercontent.com/7040974/55907832-cdae6280-5ba5-11e9-81d4-dd7391bf7e9c.gif)

This is the relevant portion of my config. 
```
    group sysinfo {
    min_length = 12
    position = 'right'
    align	= 'right'
    format = '{output}' 
    cycle = 2
     sysdata {
       min_length = 12
       position = 'right'
       on_click 1 = 'exec ksysguard'
       format = '[\?color=cpu {cpu_used_percent:02.0f} %   \uf9c4]'  
      }
    sysdata{
      min_length = 12
      position = 'right'
      on_click 1 = 'exec ksysguard'	
      format = '[\?color=mem {mem_used:02.0f}/{mem_total:02.0f} {mem_unit} \uf85a]'
    }
  sysdata {
      min_length = 12
      position = 'right'
      on_click 1 = 'exec ksysguard'	
      format = '[\?color=temp {cpu_temp:02.0f} {temp_unit}  \uf9c4]'
  }
    }

```
[py3status.conf.txt](https://github.com/ultrabug/py3status/files/3065514/py3status.conf.txt)

**Comment by lasers** [comment]
With `py3status --debug --log /tmp/py3status_bottom.log`, you can grep `method group returned` to better track times.

Also, every time the bar is hidden, it send signals to py3status to start/stop too.

Additionally, I wonder if we should add `0.1` to `cycle_time` to better offset #1650.

```diff
diff --git a/py3status/modules/group.py b/py3status/modules/group.py
index b11a70b..0a991e7 100644
--- a/py3status/modules/group.py
+++ b/py3status/modules/group.py
@@ -219,7 +219,7 @@ class Py3status:
             self.cycle = self.cycle_timeout
             if time() >= self.cycle_time:
                 self._change_active(1)
-                self.cycle_time = time() + self.cycle
+                self.cycle_time = time() + self.cycle + 0.1
 
         # time
         update_time = update_time or self.cycle or None
```

I take it this is `py3status version 3.18`? 

>Sometimes a module stays there for 4-5 seconds, sometime it switches after 1sec.

There is `MAX_NO_CONTENT_WAIT = 5` for initial content at startup too if you think you tripped that, by getting an empty content three times, but that might not be it.

Hiding bars from time to time... could be a reason why some modules stays longer or possibly that it starts counting to `2` seconds again after receiving stop/start signals? I don't know.

Anyway, you can play and `tail -f` the log to see what the conclusion is. And to try `+ 0.1` too.

**Comment by guraltsev** [comment]
So, I removed the hiding of the bar and it does not help. 

I added the log and this is what happens:

```
2019-04-10 19:45:44 INFO method group returned {'composite': [{'full_text': '    31 %   龍', 'color': '#859900', 'instance': '_anon_module_0 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:45:46 INFO method group returned {'composite': [{'full_text': ' 03/15 GiB \uf85a', 'color': '#859900', 'instance': '_anon_module_1 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:45:48 INFO method group returned {'composite': [{'full_text': ' 03/15 GiB \uf85a', 'color': '#859900', 'instance': '_anon_module_1 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:45:50 INFO method group returned {'composite': [{'full_text': '    37 °C  龍', 'color': '#859900', 'instance': '_anon_module_2 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:45:52 INFO method group returned {'composite': [{'full_text': '    31 %   龍', 'color': '#859900', 'instance': '_anon_module_0 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:45:52 INFO clearing cache for method group
2019-04-10 19:45:52 INFO method group returned {'composite': [{'full_text': '    04 %   龍', 'color': '#859900', 'instance': '_anon_module_0 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:45:54 INFO method group returned {'composite': [{'full_text': ' 03/15 GiB \uf85a', 'color': '#859900', 'instance': '_anon_module_1 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:45:56 INFO method group returned {'composite': [{'full_text': ' 03/15 GiB \uf85a', 'color': '#859900', 'instance': '_anon_module_1 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:45:58 INFO method group returned {'composite': [{'full_text': '    33 °C  龍', 'color': '#859900', 'instance': '_anon_module_2 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:46:00 INFO method group returned {'composite': [{'full_text': '    33 °C  龍', 'color': '#859900', 'instance': '_anon_module_2 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:46:02 INFO method group returned {'composite': [{'full_text': '    04 %   龍', 'color': '#859900', 'instance': '_anon_module_0 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:46:04 INFO method group returned {'composite': [{'full_text': ' 03/15 GiB \uf85a', 'color': '#859900', 'instance': '_anon_module_1 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:46:06 INFO method group returned {'composite': [{'full_text': ' 03/15 GiB \uf85a', 'color': '#859900', 'instance': '_anon_module_1 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:46:08 INFO method group returned {'composite': [{'full_text': '    33 °C  龍', 'color': '#859900', 'instance': '_anon_module_2 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:46:10 INFO method group returned {'composite': [{'full_text': '    03 %   龍', 'color': '#859900', 'instance': '_anon_module_0 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:46:12 INFO method group returned {'composite': [{'full_text': ' 03/15 GiB \uf85a', 'color': '#859900', 'instance': '_anon_module_1 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:46:14 INFO method group returned {'composite': [{'full_text': '    33 °C  龍', 'color': '#859900', 'instance': '_anon_module_2 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-10 19:46:16 INFO method group returned {'composite': [{'full_text': '    33 °C  龍', 'color': '#859900', 'instance': '_anon_module_2 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
```

my ```py3status``` version is 

```
$ py3status --version
py3status version 3.18 (python 3.7.3) on i3
```

**Comment by guraltsev** [comment]
I also tried the ```+0.1```. It does not help. It seems, based on the logs that the rotation does happen every 2 seconds but for whatever reason often it doesn't go to the next element of the group but reshows the original one.

**Comment by lasers** [comment]
Can you revert #1650 and try again?

**Comment by guraltsev** [comment]
The diff by itself does nothing. The bug is present in the same manner as before.

**Comment by guraltsev** [comment]
I reverted the commit you mentioned but it doesn't really solve the problem:

```
2019-04-11 16:51:42 INFO method group returned {'composite': [{'full_text': ' 02/15 GiB \uf85a', 'color': '#859900', 'instance': '_anon_module_1 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-11 16:51:45 INFO method group returned {'composite': [{'full_text': '    33 °C  龍', 'color': '#859900', 'instance': '_anon_module_2 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-11 16:51:48 INFO method group returned {'composite': [{'full_text': '    04 %   龍', 'color': '#859900', 'instance': '_anon_module_0 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-11 16:51:51 INFO method group returned {'composite': [{'full_text': ' 02/15 GiB \uf85a', 'color': '#859900', 'instance': '_anon_module_1 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-11 16:51:51 INFO clearing cache for method group
2019-04-11 16:51:51 INFO method group returned {'composite': [{'full_text': '    32 °C  龍', 'color': '#859900', 'instance': '_anon_module_2 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-11 16:51:53 INFO method group returned {'composite': [{'full_text': '    32 °C  龍', 'color': '#859900', 'instance': '_anon_module_2 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-11 16:51:54 INFO clearing cache for method group
2019-04-11 16:51:54 INFO method group returned {'composite': [{'full_text': '    01 %   龍', 'color': '#859900', 'instance': '_anon_module_0 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-11 16:51:57 INFO method group returned {'composite': [{'full_text': '    01 %   龍', 'color': '#859900', 'instance': '_anon_module_0 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-11 16:52:00 INFO method group returned {'composite': [{'full_text': ' 02/15 GiB \uf85a', 'color': '#859900', 'instance': '_anon_module_1 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-11 16:52:03 INFO method group returned {'composite': [{'full_text': '    33 °C  龍', 'color': '#859900', 'instance': '_anon_module_2 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-11 16:52:06 INFO method group returned {'composite': [{'full_text': '    02 %   龍', 'color': '#859900', 'instance': '_anon_module_0 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-11 16:52:09 INFO method group returned {'composite': [{'full_text': ' 02/15 GiB \uf85a', 'color': '#859900', 'instance': '_anon_module_1 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-11 16:52:09 INFO clearing cache for method group
2019-04-11 16:52:09 INFO method group returned {'composite': [{'full_text': '    32 °C  龍', 'color': '#859900', 'instance': '_anon_module_2 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
2019-04-11 16:52:12 INFO method group returned {'composite': [{'full_text': '    32 °C  龍', 'color': '#859900', 'instance': '_anon_module_2 0', 'name': 'sysdata'}], 'instance': 'sysinfo', 'name': 'group'} 
```

Practically the current version situation updates the group regularly (every 2 secs) but the cycle is wrong in the sense sometimes the same group element gets displayed several times. 

Reverting the commit causes the update interval to be irregular: instead of constant 2 second intervals you get random 2-4 second updates.

**Comment by guraltsev** [comment]
Putting *both* the diff and reverting the commit also does not help. BTW, I deleted 
` /usr/lib/python3.7/site-packages/py3status/__pycache__`
is the file being cached anywhere else? I obviously restart i3wm and the bar to test.

**Comment by lasers** [comment]
If you want, you can uninstall py3status and set up a development environment. https://py3status.readthedocs.io/en/latest/contributing.html#setting-up-a-development-environment

Also, to make a config for `group sysinfo` without other modules so you can parse logs better. I'm using this diff to see if anything different turns up. It seems to be on point, but reverting seems to help when we have many other non-group modules in play too.
```diff
diff --git a/py3status/modules/group.py b/py3status/modules/group.py
index b11a70b..3d0693a 100644
--- a/py3status/modules/group.py
+++ b/py3status/modules/group.py
@@ -217,7 +217,11 @@ class Py3status:
         # keep cycling if defined and no urgent
         if self.cycle_timeout and not urgent:
             self.cycle = self.cycle_timeout
-            if time() >= self.cycle_time:
+            _time = time()
+            _cycle_time = self.cycle_time
+            _answer = _time >= _cycle_time
+            self.py3.log("{} vs {} = {}".format(_time, _cycle_time, _answer))
+            if _answer:
                 self._change_active(1)
                 self.cycle_time = time() + self.cycle
 
@@ -225,6 +229,7 @@ class Py3status:
         update_time = update_time or self.cycle or None
         if update_time is not None:
             cached_until = self.py3.time_in(self.cycle)
+            # cached_until = self.py3.time_in(sync_to=self.cycle)
         else:
             cached_until = self.py3.CACHE_FOREVER
 
diff --git a/py3status/py3.py b/py3status/py3.py
index 1c30fa1..773505e 100644
--- a/py3status/py3.py
+++ b/py3status/py3.py
@@ -666,6 +666,8 @@ class Py3:
         if sync_to:
             requested = (requested + sync_to) - (requested % sync_to)
 
+        self.log(requested)
+
         return requested + offset
 
     def format_contains(self, format_string, names):

```

**Comment by lasers** [comment]
@guraltsev Can you try this branch? https://github.com/ultrabug/py3status/tree/group-cycle-time. Thx.

**Comment by ultrabug** [comment]
Fix will be present in 3.19 ! cheers
